### PR TITLE
JAMES-2763 ElasticSearch startup check

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchIndexesCreator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.apache.james.backends.es.ElasticSearchConfiguration;
+import org.apache.james.lifecycle.api.Startable;
+import org.apache.james.mailbox.elasticsearch.ElasticSearchMailboxConfiguration;
+import org.apache.james.mailbox.elasticsearch.MailboxIndexCreationUtil;
+import org.apache.james.quota.search.elasticsearch.ElasticSearchQuotaConfiguration;
+import org.apache.james.quota.search.elasticsearch.QuotaSearchIndexCreationUtil;
+import org.elasticsearch.client.RestHighLevelClient;
+
+class ElasticSearchIndexesCreator implements Startable {
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchIndexesCreator.java]: Maybe we should avoid having quota-search - which might be a separated module - sharing code with mailbox-elasticsearch anymore.

I'd be in favor of having 2 classes.

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java [hunk 2]
```diff
@@ -28,24 +28,53 @@
...
 import feign.Feign;
 import feign.Logger;
 import feign.RequestLine;
 import feign.Response;
+import feign.jackson.JacksonDecoder;
 import feign.slf4j.Slf4jLogger;
 
 public class DockerElasticSearch {
 
+    public static class About {
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java]: Again, why is this API needed at all?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            "ES version(%s) is not compatible with the recommendation(%s)",
+            esVersion.toString(),
+            RECOMMENDED_ES_VERSION.toString());
+    }
+
+    private final RestHighLevelClient client;
+
+    @Inject
+    private ElasticSearchStartUpCheck(ElasticSearchConfiguration configuration) {
+        this.client = ClientProviderImpl.fromHosts(configuration.getHosts(), configuration.getClusterName())
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: We should inject the client.

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/JamesWithNonCompatibleElasticSearchServerTest.java [hunk 4]
```diff
@@ -58,7 +60,14 @@ static void afterAll() {
...
     void jamesShouldStopWhenStartingWithANonCompatibleElasticSearchServer(GuiceJamesServer server) throws Exception {
         assertThatThrownBy(server::start)
-            .isInstanceOf(ElasticsearchStatusException.class);
+            .isInstanceOfSatisfying(
+                StartUpChecksPerformer.StartUpChecksException.class,
+                ex -> assertThat(ex.getBadChecks())
+                    .contains(CheckResult.builder()
+                        .checkName(ElasticSearchStartUpCheck.CHECK_NAME)
+                        .resultType(StartUpCheck.ResultType.BAD)
+                        .description(ElasticSearchStartUpCheck.notCompatibleWith(dockerES2.esVersion()))
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/JamesWithNonCompatibleElasticSearchServerTest.java]: Write the plain String here please - that would be waay simpler to write and understand.

**Comment by Arsnael** [comment]
Read it

**Comment by Arsnael** [comment]
```
[950b89916c76b8af185ef711fbdfe9cce3cfc810] [ERROR] reasonableLongTermShouldNotBeIgnored  Time elapsed: 5.22 s  <<< ERROR!
[950b89916c76b8af185ef711fbdfe9cce3cfc810] java.lang.NoClassDefFoundError: feign/Logger
[950b89916c76b8af185ef711fbdfe9cce3cfc810] Caused by: java.lang.ClassNotFoundException: feign.Logger
[950b89916c76b8af185ef711fbdfe9cce3cfc810] 
[950b89916c76b8af185ef711fbdfe9cce3cfc810] [ERROR] fieldsExceedingLuceneLimitShouldNotBeIgnored  Time elapsed: 1.704 s  <<< ERROR!
[950b89916c76b8af185ef711fbdfe9cce3cfc810] java.lang.NoClassDefFoundError: feign/Logger
[950b89916c76b8af185ef711fbdfe9cce3cfc810] Caused by: java.lang.ClassNotFoundException: feign.Logger
[950b89916c76b8af185ef711fbdfe9cce3cfc810] 
[950b89916c76b8af185ef711fbdfe9cce3cfc810] [ERROR] tooLongTermsShouldNotMakeIndexingFail  Time elapsed: 1.421 s  <<< ERROR!
[950b89916c76b8af185ef711fbdfe9cce3cfc810] java.lang.NoClassDefFoundError: feign/Logger
[950b89916c76b8af185ef711fbdfe9cce3cfc810] Caused by: java.lang.ClassNotFoundException: feign.Logger

[...]
```
Many like these, fix please

**Comment by chibenwa** [comment]
> Many like these, fix please

Easy fix: get rid of feign ^^ :trollface:

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchClientModule.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Provides
+    @Singleton
+    protected RestHighLevelClient provideClient(ElasticSearchConfiguration configuration) {
+
+        Duration waitDelay = Duration.ofMillis(configuration.getMinDelay());
+        return Mono.fromCallable(() -> connectToCluster(configuration))
+            .doOnError(e -> LOGGER.warn("Error establishing ElasticSearch connection. Next retry scheduled in {}",
+                DurationFormatUtils.formatDurationWords(waitDelay.toMillis(), true, true),
```
Comment (blackheaven) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchClientModule.java]: Remove the extra CR

**Comment by chibenwa** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[f2c70488c85cd07c6d89d10ddef91d6d218c1275] 2019-05-29 05:39:24,945 ERROR [Thread-477] - [spark.Spark]- ignite failed
[f2c70488c85cd07c6d89d10ddef91d6d218c1275] java.net.BindException: Address already in use

Webadmin intergratiion test module
```
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**